### PR TITLE
Detect and handle breaking changes in schemas

### PR DESCRIPTION
**Pull request by dominiklohmann**
###  :notebook_with_decorative_cover: Description

When changing schemas such that types are no longer a superset of previous versions of the same type, VAST may end up in a state where it cannot import events of that type.

This change makes the type registry ~forget about incompatible versions of types automatically when a new and incompatible type is registered.~ warn when this happens and ensures that newly registered types get priority over older types when importing data.

###  :memo: Checklist

- [x] All user-facing changes have changelog entries.
- [x] The changes are reflected on [docs.tenzir.com/vast](https://docs.tenzir.com/vast), if necessary.
- [x] The PR description contains instructions for the reviewer, if necessary.

### :dart: Review Instructions

Commit-by-commit.

### CODE REVIEW

File: schema/test.schema
Comment (mavam) [inline schema/test.schema:11]: I feel that testing the feature by introducing a mismatch between the schemas we ship and the hard-coded ones is difficult to argue. Instead, you could the custom schema in the integration tests, no?

File: libvast/src/system/type_registry.cpp
Comment (mavam) [inline libvast/src/system/type_registry.cpp:116]: Would anything prevent against beefing up the function `congruent` with that logic? Then it could also be truly recursive.

File: libvast/src/system/type_registry.cpp
Comment (mavam) [inline libvast/src/system/type_registry.cpp:122]:
```suggestion
    VAST_WARNING(self, "detected an incompatible version of",
```

Up to you, but I'd see this as something beyond VERBOSE.

**Review by mavam — CHANGES_REQUESTED** [review]
I wonder whether there's a good way to unit-test this?!

File: schema/test.schema
Comment (dominiklohmann) [reply schema/test.schema:11]: This is a bugfix. There's a second field above also called `s` that has a different type. This triggered the newly introduced mechanism on every start of the type registry.

File: libvast/src/system/type_registry.cpp
Comment (dominiklohmann) [reply libvast/src/system/type_registry.cpp:116]: I'd rather duplicate the logic and call it `compatible`. This is not about congruency. Note that as is this expects flattened types.

File: libvast/src/system/type_registry.cpp
Comment (dominiklohmann) [reply libvast/src/system/type_registry.cpp:122]: Fine by me.

**Comment by dominiklohmann** [comment]
> I wonder whether there's a good way to unit-test this?!

I've added a unit test as a separate commit. It isn't beautiful, but it works.

File: libvast/test/system/type_registry.cpp
Comment (mavam) [inline libvast/test/system/type_registry.cpp:71]: In what sense incompatible?

Comment (dominiklohmann) [reply libvast/test/system/type_registry.cpp]: I've expanded on this a bit with the most recent commit (c6ce834)

File: libvast/vast/type.hpp
Comment (mavam) [inline libvast/vast/type.hpp:827]: Please explain how you define "superset".

Comment (mavam) [reply libvast/vast/type.hpp:827]: Also, I would flip the argument and call it `is_subset`, which is the more common way to look at a type: one type is contained (= subset) within another.

File: libvast/src/type.cpp
Comment (mavam) [inline libvast/src/type.cpp:709]:
```suggestion
  auto sup = caf::get_if<record_type>(&y);
```

For consistency.

File: libvast/src/type.cpp
Comment (mavam) [inline libvast/src/type.cpp:715]: The "no new fields were added" confuses me. It's about checking difference between two given types, no?

File: libvast/src/type.cpp
Comment (mavam) [inline libvast/src/type.cpp:716]:
```suggestion
  for (auto& field : sub->fields) {
```

File: libvast/src/type.cpp
Comment (mavam) [inline libvast/src/type.cpp:718]:
```suggestion
      // Perform the check recursively to support nested record types.
```

File: libvast/src/type.cpp
Comment (mavam) [inline libvast/src/type.cpp:726]:
```suggestion
```

Unfortunately this not adding much insight, so we can ditch it.

File: libvast/src/system/type_registry.cpp
Comment (mavam) [inline libvast/src/system/type_registry.cpp:104]:
```suggestion
                 "and deletes existing versions");
```

**Review by mavam — DISMISSED** [review]
The logic looks sounds to me. Just small stuff now before merging.

File: libvast/src/type.cpp
Comment (dominiklohmann) [reply libvast/src/type.cpp:716]: We should use `const auto&` or `auto&&` for this, but never `auto&`.

```suggestion
  for (const auto&& field : sub->fields) {
```

File: libvast/src/type.cpp
Comment (dominiklohmann) [reply libvast/src/type.cpp:709]: Not feeling this one, nobody ever said supset.

File: libvast/src/type.cpp
Comment (dominiklohmann) [reply libvast/src/type.cpp:715]: Ah yes, this comment moved over from when this was inline in the type registry, where we had a new and multiple old types.

Comment (dominiklohmann) [reply libvast/src/type.cpp:715]:
```suggestion
  // Check whether all fields of the subset exist in the superset.
```

**Comment by dominiklohmann** [comment]
This should be reviewed by @tobim after our Slack call yesterday on the matter. The last commit is the most important one here.

File: libvast/vast/system/source.hpp
Comment (tobim) [inline libvast/vast/system/source.hpp:168]: This and the `clear()` in the next line have the effect that a modified `*.schema` file that is passed to the `--schema-file=` option will not be considered. I know this is not a result of this change, but I just stumbled over this while testing this change.

File: libvast/src/system/type_registry.cpp
Comment (tobim) [inline libvast/src/system/type_registry.cpp:101]: This can be improved so that `is_subset()` is only called if the new layout was successfully inserted.
```suggestion
  // Insert into the existing bucket.
  auto [hint, success] = old_layouts.insert(std::move(new_layout));
  if (success) {
    if (hint != old_layouts.begin()) {
      if (!is_subset(*old_layouts.begin(), new_layout))
        VAST_WARNING(self, "detected an incompatible version of",
                     new_layout.name());
    }
    VAST_DEBUG(self, "registered", hint->name());
  }
```

**Review by tobim — CHANGES_REQUESTED** [review]

File: libvast/src/system/type_registry.cpp
Comment (dominiklohmann) [inline libvast/src/system/type_registry.cpp:114]: This is way too noisy on start with an empty db. We originally had this on verbose, and moved it to debug just recently. I'd say move the above "detected a layout change" from verbose to info and move this to debug again.

File: libvast/vast/system/source.hpp
Comment (dominiklohmann) [inline libvast/vast/system/source.hpp:170]: Can you drop this commit? It is entirely unrelated from my PR and we said yesterday that we want to do this in a separate PR.

Not quite sure why you've just pushed this on top, because we were already done reviewing the logic and now there needs to be yet another reviewer.

File: libvast/src/system/type_registry.cpp
Comment (tobim) [reply libvast/src/system/type_registry.cpp:114]: Ah, I disabled the loading of the on-disk types during testing. Should have done a final check...

File: libvast/vast/system/source.hpp
Comment (tobim) [reply libvast/vast/system/source.hpp:170]: I added this because it is **actually required** to fix the reported issue. If we don't give preference to the local schema at the source there is no way it can ever make its way into the type registry without a restart of the node.

**Review by tobim — APPROVED** [review]

**Pull request merged**
